cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll

```diff
@@ -83,6 +83,11 @@ class PureStrFunction extends AliasFunction, ArrayFunction, TaintFunction, SideE
   override predicate hasOnlySpecificReadSideEffects() { none() }
 
   override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  override predicate hasSpecificReadSideEffect(ParameterIndex i, boolean buffer) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType and
+    buffer = true
+  }
```

Comment:
If I'm understanding correctly, the information encoded here is exactly the same as the `hasArrayInput` predicate for the `ArrayFunction` model.  Is there a way we could safely have one deduce from the other?

Reply:
I've put it on our team meeting agenda to discuss how we can make it easier to get this modelling right in the future. Unfortunately I don't expect we'll have time to discuss it this week.

Reply:
I think it will also be tricky to get right, to be honest.

Reply:
We had a brief discussion about this after the meeting, one idea that came up was to have `ArrayFunction` model only the buffer sizes, not their input / output status - and point people to `SideEffectFunction` for that (which possibly needs explaining a bit more clearly as the read side effects in particular are not actually 'side effects' in the traditional sense).